icu4c/source/i18n/nfsubs.cpp

```diff
@@ -103,7 +103,25 @@ class MultiplierSubstitution : public NFSubstitution {
     }
 
     virtual double transformNumber(double number) const override {
-        if (getRuleSet()) {
+        bool doFloor = getRuleSet() != nullptr;
+        if (!doFloor) {
+            // This is a HACK that partially addresses ICU-22313.  The original code wanted us to do
+            // floor() on the result if we were passing it to another rule set, but not if we were passing
+            // it to a DecimalFormat.  But the DurationRules rule set has multiplier substitutions where
+            // we DO want to do the floor() operation.  What we REALLY want is to do floor() any time
+            // the owning rule also has a ModulusSubsitution, but we don't have access to that information
+            // here, so instead we're doing a floor() any time the DecimalFormat has maxFracDigits equal to
+            // 0.  This seems to work with our existing rule sets, but could be a problem in the future,
+            // but the "real" fix for DurationRules isn't worth doing, since we're deprecating DurationRules
+            // anyway.  This is enough to keep it from being egregiously wrong, without obvious side
+            // effects.     --rtg 8/16/23
+            const DecimalFormat* decimalFormat = getNumberFormat();
+            if (decimalFormat == nullptr || decimalFormat->getMaximumFractionDigits() == 0) {
+                doFloor = true;
+            }
+        }
+        
+        if (doFloor) {
```

Comment:
For discussion, here are the original rules for English.
```
%with-words:
0: 0 seconds; 1 second; =0= seconds;
60/60: <%%min<[, >>];
3600/60: <%%hr<[, >>>];
%%min:
0: 0 minutes; 1 minute; =0= minutes;
%%hr:
0: 0 hours; 1 hour; =0= hours;
%in-numerals:
0: =0= sec.;
60: =%%min-sec=;
3600: =%%hr-min-sec=;
%%min-sec:
0: :=00=;
60/60: <0<>>;
%%hr-min-sec:
0: :=00=;
60/60: <00<>>;
3600/60: <#,##0<:>>>;
%duration:
0: =%in-numerals=;
%%lenient-parse:
& ':' = '.' = ' ' = '-';
```
As I recall, the reason why this issue happens is because all of the other RBNF rules won't trigger this issue. Here's why.

1. Those rules have an `x.x` rule. They will either do the math to use the words, or they will format the invalid fraction rule with `=#,##0.#=` in the scenario like ordinals. Any results from fractional ordinals are considered undefined because it's bad input. So we don't test it for accuracy beyond showing that something is formatted.
2. When a number is formatted, it's typically using an integer rule like the following `=#,##0=` for really large values instead of a floating point. Usually this is hit well beyond the precision of the mantissa of a little over 9 quadrillion. For the vetted RBNF languages, this rule is typically hit around 1 quintillion - 1. A maximum signed 64-bit integer is around 9.22 quintillion.

There hasn't been a tested scenario where you're formatting a floating point number instead of words, **and** applying a transformation on it without an x.x rule. So we haven't used `=00=` in a non-integer situation before in the other rules.

So the reported rounding is bad, but none of the **supported** rules really hit it. It's yet another reason why RBNF isn't suited well for durations.